### PR TITLE
Add insert! and pushfirst!

### PR DESCRIPTION
**Pull request by bkamins**
Fixes https://github.com/JuliaData/DataFrames.jl/issues/2936

@nalimilan - this PR + https://github.com/JuliaData/DataFrames.jl/pull/3071 (small patch) will complete the row-oriented API for DataFrames.jl.

(unfortunately it is large, that is why I made earlier PRs in this sequence separately)

### CODE REVIEW

**Comment by bkamins** [comment]
In the PR I fall back with `push!` and `pushfirst!` to `insert!`. It will have a minimal performance impact, but it should be negligible (and the benefit is that I avoid code duplication).

File: src/dataframe/dataframe.jl [hunk 1]
```diff
@@ -1741,78 +1769,96 @@ As a special case, if `df` has no columns and `row` is a `NamedTuple` or
...
+
+function Base.insert!(df::DataFrame, loc::Integer, row::Any; promote::Bool=false)
+    loc isa Bool && throw(ArgumentError("invalid index: $loc of type Bool"))
+    1 <= loc <= nrow(df)+1 || throw(ArgumentError("invalid index: $loc for data " *
+                                                "frame with $(nrow(df)) rows"))
     if !(row isa Union{Tuple, AbstractArray})
         # an explicit error is thrown as this was allowed in the past
-        throw(ArgumentError("`push!` does not allow passing collections of type " *
+        throw(ArgumentError("`insert!` does not allow passing collections of type " *
                             "$(typeof(row)) to be pushed into a DataFrame. Only " *
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl:1862]: Better not mention the function given that the error can happen when calling `push!` and `pushfirst!`?
```suggestion
        throw(ArgumentError("it is not allowed to insert collections of type " *
                            "$(typeof(row)) into a DataFrame. Only " *
```

File: src/dataframe/dataframe.jl [hunk 2]
```diff
@@ -1827,27 +1873,25 @@ function Base.push!(df::DataFrame, row::Any; promote::Bool=false)
...
-        for col in _columns(df)
-            current_col += 1
-            @assert length(col) == targetrows
-        end
     catch err
         #clean up partial row
         for col in _columns(df)
-            resize!(col, nrows)
+            length(col) == targetrows && deleteat!(col, loc)
+            @assert length(col) == nrows
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl:1894]: I imagine the new code could give weird results in a corner case where several columns at the end of the data frame already had the target size (but not others). Or if an unrelated error happens when adding a row to such a data frame. Then `@assert` above will fail for the first one, and we will delete one entry at position `loc` in each of the other too long columns. So at the end the data frame won't be the same as before calling `insert!`. Not sure whether we should care or not.

File: src/dataframerow/dataframerow.jl [hunk 3]
```diff
@@ -496,10 +496,22 @@ function DataFrame(dfr::DataFrameRow)
...
+                promote::Bool=(cols in [:union, :subset])) =
+    insert!(df, 1, row, cols=cols, promote=promote)
+
+@noinline pushhelper!(x, loc, r) = insert!(x, loc, x[r])
+
+function Base.insert!(df::DataFrame, loc::Integer, dfr::DataFrameRow;
+                      cols::Symbol=:setequal, promote::Bool=(cols in [:union, :subset]))
+    loc isa Bool && throw(ArgumentError("invalid index: $loc of type Bool"))
+    1 <= loc <= nrow(df)+1 || throw(ArgumentError("invalid index: $loc for data " *
+                                                "frame with $(nrow(df)) rows"))
```
Comment (nalimilan) [inline src/dataframerow/dataframerow.jl:514]:
```suggestion
    1 <= loc <= nrow(df)+1 ||
        throw(ArgumentError("invalid index: $loc for data frame with $(nrow(df)) rows"))
```

File: src/dataframerow/dataframerow.jl [hunk 4]
```diff
@@ -496,10 +496,22 @@ function DataFrame(dfr::DataFrameRow)
...
+Base.push!(df::DataFrame, row::DataFrameRow;
+           cols::Symbol=:setequal,
+           promote::Bool=(cols in [:union, :subset])) =
+    insert!(df, nrow(df)+1, row, cols=cols, promote=promote)
+Base.pushfirst!(df::DataFrame, row::DataFrameRow;
+                cols::Symbol=:setequal,
+                promote::Bool=(cols in [:union, :subset])) =
+    insert!(df, 1, row, cols=cols, promote=promote)
+
+@noinline pushhelper!(x, loc, r) = insert!(x, loc, x[r])
```
Comment (nalimilan) [inline src/dataframerow/dataframerow.jl:508]: Make signature more specific if possible?

File: test/runtests.jl [hunk 5]
```diff
@@ -21,6 +21,7 @@ my_tests = ["utils.jl",
             "data.jl",
             "index.jl",
             "dataframe.jl",
+            "push.jl",
```
Comment (nalimilan) [inline test/runtests.jl:24]: "insert.jl" or "insertion.jl" would sound like a more general name to me.

File: src/dataframe/dataframe.jl [hunk 6]
```diff
@@ -1674,12 +1689,13 @@ function Base.push!(df::DataFrame, row::Union{AbstractDict, NamedTuple};
...
-                push!(col, val)
+                insert!(col, loc, val)
             else
                 newcol = similar(col, promote_type(S, T), targetrows)
-                copyto!(newcol, 1, col, 1, nrows)
-                newcol[end] = val
                 firstindex(newcol) != 1 && _onebased_check_error()
+                copyto!(newcol, 1, col, 1, loc-1)
+                copyto!(newcol, loc+1, col, loc, nrows-loc+1)
+                newcol[loc] = val
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl:1698]: In terms of cache locality, this would be more logical. Probably doesn't matter in practice though. (Same elsewhere.)
```suggestion
                copyto!(newcol, 1, col, 1, loc-1)
                newcol[loc] = val
                copyto!(newcol, loc+1, col, loc, nrows-loc+1)
```

**Review by nalimilan — COMMENTED** [review]
Have you checked whether the performance impact is indeed small? Sounds likely given the overhead of accessing each column anyway.

File: src/dataframe/dataframe.jl (hunk 2 above)
Comment (bkamins) [reply src/dataframe/dataframe.jl:1894]: > I imagine the new code could give weird results in a corner case where several columns at the end of the data frame already had the target size (but not others).

In normal circumstances it is not possible. The reason is that in the loop above I fail fast with `@assert length(col) == nrows` check in the `try` block. Which means that we exit the loop the first time something wrong happens in processing of the data. In other words if we get an error in column e.g. 3 then columns 4 and later are not touched

> Or if an unrelated error happens when adding a row to such a data frame.

Indeed if some system-level error happens then things might be garbage. But in such a situation there is no way to recover from the error (as things must have failed on a very low level, so we anyway do not know how to recover). The code assumes that if things failed and we see a column into which a row was added then this addition happened at `loc` position.

> Then `@assert` above will fail for the first one, and we will delete one entry at position loc in each of the other too long columns.

This is not possible under "normal" work of Julia (i.e. if we do not hit some system level error). As commented above. If we have an error in column 3, later columns are not touched. The only case when they would be touched is when are aliases of e.g. column 3, but in this case we recover from error in column `3` and some later column is exactly the same as column 3 (as it is its alias).

Note that we cannot have a potential problem of overlapping memory segments that can happen in views because views are not resizable, so function will error early anyway.

---

Side note: if you compare this implementation to the `push!` implementation indeed it is more complex (and it has to be). The issue is that in `push!` we knew we have to trim all the vectors and did not have to "fail fast" on aliases. Now we need to fail fast on aliases , as otherwise indeed we would push multiple values at location `loc` and then we would not know how many to remove. In the current code we fail the first time we see something wrong in a column (so at most one incorrect insertion is made).

In short: I have thought about it when implementing the code. Things should be OK, but are not easy to reason about. We try to recover the proper state of the data frame as well as we can, but in general users should not rely on a fact that in-place operation properly recovers the state of the mutated object on error (Base Julia does not ensure this; we try, and I would say that most of the time we do it correctly).

File: test/insertion.jl [hunk 7]
```diff
@@ -0,0 +1,1253 @@
...
+    @test insert!(df, 3, df[1, :]) == DataFrame(a=[3; 1; 3; 2:3; 2], b=[4; 2; 4; 3:4; 3], c=[5; 3; 5; 4:5; 4])
+    df = DataFrame(a=1:3, b=2:4)
+    df.c = df.a
+    @test_throws AssertionError push!(df, df[2, :])
+    @test_throws AssertionError pushfirst!(df, df[2, :])
+    @test_throws AssertionError insert!(df, 2, df[2, :])
+    @test df == DataFrame(a=1:3, b=2:4, c=1:3)
+end
+
+@testset "multicolumn aliasing" begin
```
Comment (bkamins) [inline test/insertion.jl:1239]: @nalimilan - this test is made to additionally make sure that we correctly recover after error in complex multi-alias cases.

**Comment by bkamins** [comment]
> Have you checked whether the performance impact is indeed small?

Indeed the performance impact is noticeable unfortunately. I will work on it and let you know (all other issues are resolved)

**Comment by bkamins** [comment]
As usual if we go into details of performance complexity comes:
1. interestingly the problematic check was `nrow(df)`, which we should avoid if possible since it is type unstable (of course sometimes it is unavoidable)
2. I found a case when old `push!` code was allowing alias in case when in the new code we catch it (and I think it is better we catch all cases of aliasing)

Which means I ended up with custom implementations of all cases :( (thus the PR is even larger).

Here is the benchmark code:
```
function push_nt()
    df = DataFrame(rand(1, 10), :auto)
    x = copy(df[1, :]) # NamedTuple
    for i in 1:1_000_000
        push!(df, x)
    end
end

function push_vec()
    df = DataFrame(rand(1, 10), :auto)
    x = ones(10) # vector
    for i in 1:1_000_000
        push!(df, x)
    end
end

function pushfirst_nt()
    df = DataFrame(rand(1, 10), :auto)
    x = copy(df[1, :]) # NamedTuple
    for i in 1:1_000_000
        pushfirst!(df, x)
    end
end

function pushfirst_vec()
    df = DataFrame(rand(1, 10), :auto)
    x = ones(10) # vector
    for i in 1:1_000_000
        pushfirst!(df, x)
    end
end

function insert_nt()
    df = DataFrame(rand(1, 10), :auto)
    x = copy(df[1, :]) # NamedTuple
    for i in 1:1_000_000
        insert!(df, i+1, x)
    end
end

function insert_vec()
    df = DataFrame(rand(1, 10), :auto)
    x = ones(10) # vector
    for i in 1:1_000_000
        insert!(df, i+1, x)
    end
end

using BenchmarkTools
@btime push_nt();
@btime push_vec();
@btime pushfirst_nt();
@btime pushfirst_vec();
@btime insert_nt();
@btime insert_vec();
```

And the results:

THIS PR:
```
julia> @btime push_nt();
  1.721 s (53989473 allocations: 1.64 GiB)

julia> @btime push_vec();
  1.003 s (30989449 allocations: 502.43 MiB)

julia> @btime pushfirst_nt();
  2.421 s (73979273 allocations: 1.94 GiB)

julia> @btime pushfirst_vec();
  1.144 s (30989449 allocations: 502.43 MiB)

julia> @btime insert_nt();
  2.445 s (85973152 allocations: 2.12 GiB)

julia> @btime insert_vec();
  1.211 s (41983838 allocations: 670.19 MiB)
```

and before this PR (`push!` only):
```
julia> @btime push_nt();
  1.873 s (53989493 allocations: 1.64 GiB)

julia> @btime push_vec();
  1.085 s (30989469 allocations: 502.43 MiB)
```

So:
1. current `push!` is minimally faster than the previous implementation. (this is probably within measurement error)
2. it is worth to have separate `push!` implementation
3. `pushfirst!` is minimally faster than `insert!` - we could drop custom implementation of `pushfirst!`, but I kept it because it allocates less (so it will trigger GC less often apart from benchmark considerations)

File: src/dataframe/dataframe.jl (hunk 2 above)
Comment (nalimilan) [reply src/dataframe/dataframe.jl]: The situation I was thinking about is e.g. this one:
```julia
julia> df = DataFrame(x=1:3, y=1:3, z=1:3)
pus3×3 DataFrame
 Row │ x      y      z     
     │ Int64  Int64  Int64 
─────┼─────────────────────
   1 │     1      1      1
   2 │     2      2      2
   3 │     3      3      3

julia> push!(df.z, 4)
4-element Vector{Int64}:
 1
 2
 3
 4

julia> push!(df, (x=1, y="a", z=2))
┌ Error: Error adding value to column :y.
└ @ DataFrames ~/.julia/dev/DataFrames/src/dataframe/dataframe.jl:1713
ERROR: MethodError: Cannot `convert` an object of type String to an object of type Int64

julia> df
3×3 DataFrame
 Row │ x      y      z     
     │ Int64  Int64  Int64 
─────┼─────────────────────
   1 │     1      1      1
   2 │     2      2      2
   3 │     3      3      3
```
Note that the last value in column z was silently dropped, without signalling that the data frame was corrupt from the beginning. This is of course a rare situation, but the previous approach of keeping track of the index of the last column that was modified before the error happened was safer.

**Comment by nalimilan** [comment]
Ah too bad. Wouldn't there be a way to share some of the code between the three functions?

**Comment by bkamins** [comment]
> Wouldn't there be a way to share some of the code between the three functions?

OK - I will share it and let you decide if it is simpler or not.

File: src/dataframe/dataframe.jl (hunk 2 above)
Comment (bkamins) [reply src/dataframe/dataframe.jl]: > This is of course a rare situation, but the previous approach of keeping track of the index of the last column that was modified before the error happened was safer.

OK - I will make the "rollback" only affect the changed columns.

**Comment by bkamins** [comment]
@nalimilan . I have refactored code so that:
1. there is no code duplication (by using `Val` all is optimized-out so there is no performance cost of such an approach - I have run the benchmarks I checked above and they are good)
2. all "insertion" is moved into a single file (it was needed because of `DataFrameRow`, which is defined in another file)
3. I moved `append!` and `prepend!` to the same file as they logically match the `push!` and friends
4. I have made sure that we "rollback" only the columns that were updated. There is one case when we do not correctly roll-back, but it is only for corrupted data frame, and I would leave it as it is now (i.e. with incorrect rollback):
```
julia> df = DataFrame(a=1:3, b=4:6)
3×2 DataFrame
 Row │ a      b     
     │ Int64  Int64 
─────┼──────────────
   1 │     1      4
   2 │     2      5
   3 │     3      6

julia> push!(df.b, 10)
4-element Vector{Int64}:
  4
  5
  6
 10

julia> push!(df, (100, 101))
┌ Error: Error adding value to column :b.
└ @ DataFrames c:\Users\bogum\.julia\dev\DataFrames\src\dataframe\insertion.jl:680
ERROR: AssertionError: length(col) == nrows

julia> df
3×2 DataFrame
 Row │ a      b     
     │ Int64  Int64 
─────┼──────────────
   1 │     1      4
   2 │     2      5
   3 │     3      6
```
(the reason why this rollback is incorrect is that `:b` had length 4 originally and code thinks that it was 4 because of successful adding of 1 row, while it was originally corrupted on purpose; of course we could also detect this case, but I think it is obscure enough to be acceptable)

File: src/dataframe/insertion.jl [hunk 8]
```diff
@@ -0,0 +1,958 @@
...
+                    copyto!(newcol, 1, col, 1, loc-1)
+                    newcol[loc] = val
+                    copyto!(newcol, loc+1, col, loc, nrows-loc+1)
+                end
+                _columns(df)[i] = newcol
+            end
+        end
+    catch err
+        #clean up partial row
+        for col2 in _columns(df)[1:current_col]
```
Comment (nalimilan) [inline src/dataframe/insertion.jl:672]: Let's avoid a small allocation? Same elsewhere.
```suggestion
        for col2 in view(_columns(df), 1:current_col)
```
or
```suggestion
        for i in 1:current_col
            col2 = _columns(df)[i]
```

File: src/dataframe/insertion.jl [hunk 9]
```diff
@@ -0,0 +1,958 @@
...
+"""
+pushfirst!
+
+Base.pushfirst!(df::DataFrame, row::Any; promote::Bool=false) =
+    _row_inserter!(df, -1, row, Val{:pushfirst}(), promote)
+
+"""
+    insert!(df::DataFrame, loc::Integer, row::Union{Tuple, AbstractArray}; promote::Bool=false)
+    insert!(df::DataFrame, loc::Integer, row::Union{DataFrameRow, NamedTuple, AbstractDict};
+            cols::Symbol=:setequal, promote::Bool=(cols in [:union, :subset]))
```
[the remark this answers is not in the document]
Comment (bkamins) [reply src/dataframe/insertion.jl:533]: OK - I will change it to `index` as this is the Base Julia term. I initially did not want to use it as we have `Index`, but probably users are not aware of this.

File: src/dataframe/insertion.jl (hunk 8 above)
Comment (bkamins) [reply src/dataframe/insertion.jl:672]: OK - I guess it is good to change for the cleanness sake. This code is only invoked on error, so we do not care if it is fast.

Comment (nalimilan) [reply src/dataframe/insertion.jl:672]: Ah right. Maybe not worth changing then.

File: src/dataframe/insertion.jl [hunk 10]
```diff
@@ -0,0 +1,958 @@
...
+    _row_inserter!(df, -1, row, Val{:pushfirst}(), cols, promote, -1)
+
+function Base.insert!(df::DataFrame, loc::Integer, row::Union{AbstractDict, NamedTuple};
+                      cols::Symbol=:setequal,
+                      promote::Bool=(cols in [:union, :subset]))
+    loc isa Bool && throw(ArgumentError("invalid index: $loc of type Bool"))
+    1 <= loc <= nrow(df)+1 ||
+        throw(ArgumentError("invalid index: $loc for data frame with $(nrow(df)) rows"))
+    return _row_inserter!(df, loc, row, Val{:insert}(), cols, promote, -1)
+end
```
[the remark this answers is not in the document]
Comment (bkamins) [reply src/dataframe/insertion.jl:769]: They are similar but not identical. The `row::DataFrameRow` methods call `_dfr_row_inserter!` instead that does an additional check if `DataFrameRow` is coming from the `DataFrame` to which we want to add it and then use a different path.

File: src/dataframe/insertion.jl (hunk 9 above)
[the remark this answers is not in the document]
Comment (bkamins) [reply src/dataframe/insertion.jl:527]: Funny enough we use the `index` function inside one of the methods. Still - I will change it to `index` in user-facing signature.

File: src/dataframe/insertion.jl (hunk 8 above)
Comment (bkamins) [reply src/dataframe/insertion.jl]: I have changed this

**Review by nalimilan — APPROVED** [review]

**Comment by bkamins** [comment]
Thank you!
